tools/tplgfw-sanity-check.py

```diff
@@ -0,0 +1,134 @@
+            "fw_binary_header" / self.Fw_binary_header,
+            "modules" / self.Module_entry[this.fw_binary_header.module_entries],
+        )
+
+    def parse_file(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_manifest.parse_stream(fw)
+
+    def parse_fw_header(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_header.parse_stream(fw)
+
+if __name__ == "__main__":
+
+    def parse_cmdline():
+        parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter,
+            description='A Python tool to match UUID in topology and FW, then find out whether all module in topoloy can be supported by this FW.\
+ Currently it only supports IPC4 FW and IPC3 & IPC4 topology. The output includes FW version and with an error message if they are not matched.\
+ And all the unsupported modules will be printed out.\
+ \n\nexample: tplgfw-sanity-check.py -t sof-tgl-nocodec.tplg -f basefw.bin')
+
+        parser.add_argument('--version', action='version', version='%(prog)s 1.0')
+        parser.add_argument('-t', type=str, nargs='+', required=True, help="""Load topology files for check """)
+        parser.add_argument('-f', type=str, nargs='+', required=True, help="""Load firmware files for check """)
+        parser.add_argument('-v', '--verbose', action="store_true", help="Show check state")
+
+        return parser.parse_args()
+
+if __name__=="__main__":
+    def main():
+        tplgFormat = tplgtool2.TplgBinaryFormat()
+        cmd_args = parse_cmdline()
+        tplg = tplgtool2.GroupedTplg(tplgFormat.parse_file(cmd_args.t[0]))
+
+        fw_parser = Fw_Parser()
+        fw_content = fw_parser.parse_file(cmd_args.f[0])
+        print("FW release version: {0}.{1}".format(fw_content.ext_header.version_major, fw_content.ext_header.version_minor))
+
+        if fw_content.ext_header.id == '$AE1':
+            print("IPC4 supported FW\n")
+        elif fw_content.ext_header.id == 'XMan':
+            print("IPC3 supported FW\n")
+        else:
+            print("Invalid FW")
+            return -1
+
+        not_found = 0
+        for widget in tplg.widget_list:
+            name = widget["widget"]["name"]
+            widget_uuid = tplg.get_widget_uuid(widget)
+            found = False
+            if widget_uuid is not None:
+                for m in fw_content.modules:
+                    if m.uuid == widget_uuid:
+                        found = True
+                        if cmd_args.verbose:
+                            print("Widget: ", name, "is supported by ", m.name)
+                        break
+
+                if found == False:
+                    not_found += 1
+                    failed_uuid = uuid.UUID(bytes=widget_uuid)
+                    print("Error: widget:", name, failed_uuid, "is not supported by this FW")
```

Comment:
I will collect all uuids (and maybe also names) in firmware and topology to two python sets, fw_uuid_set and tplg_uuid_set.
so we can use set operations: union/difference to find those widgets both in the two sets, in one set and not in the other.